### PR TITLE
[WIP DO NOT MERGE] Handle update events

### PR DESCRIPTION
**Pull request by imcsk8**
Since the operator sdk does not natively handles update
events we have to check for differences in the ClusterIngress
structure by getting its md5 signatureits

This commit is work in progress does not handle updates on the
deployment and service elements of the ClusterIngress.

Code cleanup

### CODE REVIEW

**Comment by imcsk8** [comment]
PTAL @openshift/sig-network-edge @ironcladlou @Miciah @pravisankar @knobunc

**Comment by ironcladlou** [comment]
See [topic 7](https://docs.google.com/document/d/1XwNLgig9BKzptumC80o0Ajl_69W8IbUYQPjdvqPoOsk/edit#) for some discussion on this, as well as the [upstream controller guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md) on managing resources with `Generation`. Might be worth having a quick call with @deads2k or somebody in @openshift/sig-master for some concrete advice.

For now, I'd be happy if we just unconditionally re-built the resource and issued a `PATCH` which may no-op. Diffing is going to be a more complex issue.

**Comment by soltysh** [comment]
@ironcladlou @tnozicka will know the most about generations.

File: pkg/stub/handler.go [hunk 1]
```diff
@@ -28,7 +28,12 @@ type Handler struct {
 	manifestFactory *manifests.Factory
 }
 
+var ci_md5 map[string][16]byte
+
 func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
+	if ci_md5 == nil {
+		ci_md5 = make(map[string][16]byte)
```
Comment (pravisankar) [inline pkg/stub/handler.go]: Where is this used? probably stale code

Comment (pravisankar) [reply pkg/stub/handler.go]: Difference in MD5 checksum of the object to denote change in custom resource may not be useful in our case.
Once we populate status field, status is updated periodically and md5 checksum on the object will change even though there is no change in CR spec.
I think we should store old spec (as annotation or something mechanism) and then compare with new spec to determine if update is needed. This mechanism will also provide what exactly has changed and can react accordingly.
We could get more pointers from `oc apply` implementation.

File: pkg/stub/handler.go [hunk 2]
```diff
@@ -97,7 +102,13 @@ func (h *Handler) syncIngressUpdate(ci *ingressv1alpha1.ClusterIngress) error {
 	err = sdk.Create(ds)
 	if err == nil {
 		logrus.Infof("created router daemonset %s/%s", ds.Namespace, ds.Name)
-	} else if !errors.IsAlreadyExists(err) {
+	} else if errors.IsAlreadyExists(err) {
+		err = sdk.Update(ds)
```
Comment (pravisankar) [inline pkg/stub/handler.go]: How do we know this is an update event and not resync event? (may be we want to update anyway during resync?)

File: pkg/stub/handler.go [hunk 3]
```diff
@@ -117,7 +128,13 @@ func (h *Handler) syncIngressUpdate(ci *ingressv1alpha1.ClusterIngress) error {
 			err = sdk.Create(service)
 			if err == nil {
 				logrus.Infof("created router service %s/%s", service.Namespace, service.Name)
-			} else if !errors.IsAlreadyExists(err) {
+			} else if errors.IsAlreadyExists(err) {
+				err = sdk.Update(service)
```
Comment (pravisankar) [inline pkg/stub/handler.go]: I don't think we need to update service, none of the spec changes in CR will change service object.

File: pkg/stub/handler.go (hunk 2 above)
Comment (imcsk8) [reply pkg/stub/handler.go]: When @Miciah PR gets merged we will know when there's an update. I think this pattern is good for now.

File: pkg/stub/handler.go (hunk 3 above)
Comment (imcsk8) [reply pkg/stub/handler.go]: you're right, removing this code

**Comment by ironcladlou** [comment]
Needs redesigned per https://github.com/openshift/cluster-ingress-operator/pull/49; should probably close this one

**Pull request closed without merging**